### PR TITLE
State management refactor to prepare for continuous state

### PR DESCRIPTION
**Pull request by mitchellh**
This PR refactors state management into the `state` package (new). The goal of this refactor was twofold: prepare for continous state saving (to be explained in another PR, since this PR doesn't do it) and to clean up a lot of the special casing around remote state, with a heavy lean on the former as motivation. Continuous state was turning out to be really nasty without treating remote/local state as similar.

As usual, tests weren't changed, only added. 

Features of this PR:

  * New package: `state` - Contains the interfaces for state storage and management.

  * New package: `state/remote` - Contains the interface implementation for `state.State` that uses a remote driver.

  * Removed package: `remote` - Is now part of `state/remote`

  * **Backwards incompatible change:** Remote driver configuration is now done via `-backend-config=k=v` flags on the `remote` and `init` flag. This makes it so that new remote drivers can be introduced without new weird CLI flags.

As a result of this PR, both local and remote state storage an implementation of the `state.State` interface. As a result, the `command` package no longer has to juggle a lot of `if remote { } else { }` logic, since its all just one interface.

Continuous state needed this because with remote state we want to only really update a local cache instead of writing to the remote driver on every save (expensive). But then we want to merge the local cache and the remote state no startup. Complicated in the current way, easy with this (just another `state.State` implementation).

Docs still have to be updated, but wanted to open this up.

### CODE REVIEW

File: command/command_test.go [hunk 1]
```diff
@@ -178,17 +178,20 @@ func testTempDir(t *testing.T) string {
 // testCwdDir is used to change the current working directory
 // into a test directory that should be remoted after
 func testCwd(t *testing.T) (string, string) {
-	tmp, err := ioutil.TempDir("", "remote")
+	tmp, err := ioutil.TempDir("", "tf")
 	if err != nil {
 		t.Fatalf("err: %v", err)
 	}
+
```
Comment (phinze) [inline command/command_test.go]: _@mitchellh wuz here_ :wink:

File: command/state.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+)
+
+// StateOpts are options to get the state for a command.
+type StateOpts struct {
+	// LocalPath is the path where the state is stored locally.
+	//
+	// LocalPathOut is the path where the local state will be saved. If this
+	// isn't set, it will be saved back to LocalPath.
+	LocalPath    string
+	LocalPathOut string
```
Comment (phinze) [inline command/state.go]: When does `LocalPathOut != LocalPath`?

Comment (mitchellh) [inline command/state.go]: Various commands take both a `-state` and `-state-out` parameter.

File: state/remote/state_test.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	}
+
+	state.TestState(t, s)
+}
+
+func TestState_impl(t *testing.T) {
+	var _ state.StateReader = new(State)
+	var _ state.StateWriter = new(State)
+	var _ state.StatePersister = new(State)
+	var _ state.StateRefresher = new(State)
```
Comment (phinze) [inline state/remote/state_test.go]: This is a cool pattern. Adding to my gopher-notes. :)

**Comment by phinze** [comment]
Alright, nice. This all LGTM.

File: state/remote/state_test.go (hunk 3 above)
Comment (mitchellh) [inline state/remote/state_test.go]: Yeah, it fails at compile time too which is nice.

**Pull request merged**